Feature/add access control to scavenger hunt nft

Michael-Nwachukwu opened the pull request:
# 🚀 Scavenger-Hunt Pull Request

- [x] Closes #124
- [x] Added tests (if necessary)
- [x] Run tests
- [x] Run formatting
- [x] Evidence attached
- [x] Commented the code

---

### 📌 Type of Change

- [ ] Documentation (updates to README, docs, or comments)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

---

## 📝 Changes description

- Added access control from openzeppelin 
- Configured functions for grant_minter_role, revoke_minter_role and has_minter_role. 
- Added command to grant Scavenge_hunt contract the Minter role on deployment in constructor
- Added necessary comments

Michael-Nwachukwu commented:
Hi @manlikeHB please find and review PR. Let me know of any modifications.

manlikeHB commented:
Amazing work! Add unit tests.

Michael-Nwachukwu commented:
Hi @manlikeHB i just updated. I modified already existing tests for mint functions to be called by the Scavenger_hunt contract and also added additional tests to ensure that 
1. only Scavenger_hunt contract can call mint function. 
2. Scavenger_hunt contract has minter role by default.

##Evidence 📸
<img width="1440" alt="Screenshot 2025-02-28 at 7 26 05 PM" src="https://github.com/user-attachments/assets/2d84a6b8-312e-4697-b7c9-6f67efb22315" />

manlikeHB commented on onchain/tests/test_scavenger_hunt_nft.cairo:
```diff
@@ -170,3 +225,43 @@ fn test_non_existent_badge() {
+
+    let recipient = deploy_mock_receiver();
+
+    // Set caller address to scavenger_hunt_address to simulate call from ScavengerHunt contract
```
```suggestion
    // Set caller address to unauthorized_address to simulate call from an unauthorized address
```

Michael-Nwachukwu replied:
Hi @manlikeHB just to be clear. 
I have a separate test that sets caller address to unauthorized_address to simulate call from an unauthorized address. The comment you highlighted is in fact a test case to simulate caller from an authorized address.

Michael-Nwachukwu replied:
Oh i get it now @manlikeHB. Thanks that was a typo.

manlikeHB commented on onchain/src/scavenger_hunt_nft.cairo:
```diff
@@ -62,12 +90,28 @@ mod ScavengerHuntNFT {
+        }
+        
+        // Revoke minter role from a contract or address
+        fn revoke_minter_role(ref self: ContractState, account: ContractAddress) {
```
add test for this too.

manlikeHB commented:
@Michael-Nwachukwu format your code when done.

Michael-Nwachukwu commented:
Hi @manlikeHB Done

Done! 

I added a function to grant the admin role and the minter role to scavenger_hunt_nft 
I also added extra test to grant minter role, mint and nft and ensure it mints, revokes the role and try to mint again and ensure that if fails

<img width="1440" alt="Screenshot 2025-02-28 at 11 46 23 PM" src="https://github.com/user-attachments/assets/90c9c746-18a4-42e0-b169-801abf8b7331" />

manlikeHB commented:
Thank you, LGTM!
